dnf/transaction_sr.py

```diff
@@ -0,0 +1,508 @@
+            self._base.goal.install(select=sltr, optional=self._base.conf.skip_broken)
+        elif action == "Reinstall":
+            query = query.available()
+
+            if not query:
+                self._raise_or_warn(self._skip_unavailable,
+                    _('Rpm nevra "{nevra}" not available in repositories for action "{action}".')
+                    .format(nevra=nevra, action=action))
+                return
+
+            sltr = dnf.selector.Selector(self._base.sack).set(pkg=query)
+            self._base.goal.install(select=sltr, optional=self._base.conf.skip_broken)
+        elif action in ("Upgraded", "Downgraded", "Reinstalled", "Removed", "Obsoleted"):
+            query = query.installed()
+
+            if not query:
+                # if ignore_installed is True, ignore the version of the package
+                # being removed and remove any NA
+                if action == "Removed" and self._ignore_installed:
+                    query_na = query_na.installed()
+
+                    if not query_na:
+                        self._warnings.append(
+                            _('Rpm "{na}" not installed for action "{action}".').format(na=na, action=action))
+                        return
+
+                    self._base.goal.erase(query_na[0], clean_deps=False)
+
+                    return
+
+                self._raise_or_warn(self._ignore_installed,
+                    _('Rpm nevra "{nevra}" not installed for action "{action}".').format(nevra=nevra, action=action))
+
+                return
+
+            # erasing the original version (the reverse part of an action like
+            # e.g. upgrade) is more robust, but we can't do it if
+            # skip_unavailable is True, because if the forward part of the
+            # action is skipped, we would simply remove the package here
+            if not self._skip_unavailable or action == "Removed":
+                self._base.goal.erase(query[0], clean_deps=False)
+        elif action == "Reason Change":
+            self._base.history.set_reason(query[0], reason)
+        else:
+            raise TransactionError(
+                _('Unexpected value of package action "{action}" for rpm nevra "{nevra}".')
+                    .format(action=action, nevra=nevra)
+            )
+
+    def get_data(self):
+        """
+        :returns: the loaded data of the transaction
+        """
+
+        return self._replay_data
+
+    def get_warnings(self):
+        """
+        :returns: an array of warnings gathered during the transaction replay
+        """
+
+        return self._warnings
+
+    def run(self):
+        """
+        Replays the transaction.
+        """
+
+        # Override configuration options that affect how the transaction is resolved
+        # TODO if used through API, we shouldn't change the global state here?
+        self._base.conf.clean_requirements_on_remove = False
+        self._base.conf.install_weak_deps = False
```

Comment:
Yes, but then if someone uses the functionality through the API, these won't be set and the behavior will be different and actually wrong, won't it?

Reply:
Correct, but the change of configuration is really hidden. Also please check the code of dnf or plugins and you will see that the configuration is modified in configure() section. Please take it as a best practise in the project.

Reply:
I'm not sure I understand, are you saying I should put the configuration change into `Command.configure()` even though that will make usage of transaction store/replay behave incorrectly through API?

Reply:
Incorrectly is not a correct word, but it will result in different behaviour. And I suppose to move it into history command.configure() when replay command is used.

Reply:
Different than correct unfortunately means incorrect. I'm moving it according to your suggestion...

Reply:
Thanks